ntoskrnl/io/iomgr/deviface.c

```diff
@@ -30,185 +31,450 @@ static PWCHAR BaseKeyString = L"\\Registry\\Machine\\System\\CurrentControlSet\\
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Fill it by an appropriate part of the link without prefix */
+    RtlCopyUnicodeString(&DeviceString, &LinkNameNoPrefix);
+
+    /* Allocate a munged path string */
+    MungedStringReal.Length = 0;
+    MungedStringReal.MaximumLength = MungedPrefix.Length + DeviceString.Length;
+    MungedStringReal.Buffer = ExAllocatePoolWithTag(NonPagedPool, MungedStringReal.MaximumLength, TAG_IO);
+    if (!MungedStringReal.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Fill it by munged prefix and munged string */
+    Status = RtlAppendUnicodeStringToString(&MungedStringReal, &MungedPrefix);
     if (!NT_SUCCESS(Status))
     {
-        DPRINT1("Failed to open %wZ%wZ (%x)\n", &BaseKeyU, &GuidString, Status);
-        goto cleanup;
+        DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+        return Status;
     }
 
-    SubKeyName.MaximumLength = SymbolicLinkName->Length + sizeof(WCHAR);
-    SubKeyName.Length = 0;
-    SubKeyName.Buffer = ExAllocatePool(PagedPool, SubKeyName.MaximumLength);
-    if (!SubKeyName.Buffer)
+    Status = RtlAppendUnicodeStringToString(&MungedStringReal, &DeviceString);
+    if (!NT_SUCCESS(Status))
     {
-        Status = STATUS_INSUFFICIENT_RESOURCES;
-        goto cleanup;
+        DPRINT1("RtlAppendUnicodeStringToString() failed with status 0x%08lx\n", Status);
+        return Status;
     }
 
-    RtlAppendUnicodeStringToString(&SubKeyName,
-                                   SymbolicLinkName);
+    DPRINT("Munged string is '%wZ'\n", &MungedStringReal);
+    DPRINT("GUID string is '%wZ'\n", &GuidStringReal);
+    DPRINT("Reference string is '%wZ'\n", &ReferenceStringReal);
 
-    SubKeyName.Buffer[SubKeyName.Length / sizeof(WCHAR)] = UNICODE_NULL;
+    /* Store received parts if the parameters are not null */
+    if (PrefixString != NULL)
+    {
+        if (RtlPrefixUnicodeString(&KernelModePrefix, SymbolicLinkName, FALSE))
+            *PrefixString = KernelModePrefix;
+        else if (RtlPrefixUnicodeString(&UserModePrefix, SymbolicLinkName, FALSE))
+            *PrefixString = UserModePrefix;
+    }
+
+    if (MungedString != NULL)
+        *MungedString = MungedStringReal;
+
+    if (GuidString != NULL)
+        *GuidString = GuidStringReal;
+
+    if (ReferenceString != NULL)
+        *ReferenceString = ReferenceStringReal;
+
+    if (ReferenceStringPresent != NULL)
+        *ReferenceStringPresent = ReferenceStringReal.Length == 1 ? FALSE : TRUE;
+
+    if (InterfaceClassGuid != NULL)
+    {
+        /* Convert GUID string into a GUID and store it also */
+        Status = RtlGUIDFromString(&GuidStringReal, InterfaceClassGuid);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("RtlGUIDFromString() failed with status 0x%08lx\n", Status);
+            return Status;
```

Comment:
You've already allocated the output buffer for `MungedString` at this point. Expecting the caller to free a returned string if the function returns failure is poor API design -- you should free it here (or move this possible failure path to before the allocation happens).

Reply:
Done.